index/scorch/persister.go

```diff
@@ -871,24 +886,79 @@ func (s *Scorch) removeOldData() {
 // rollback'ability.
 var NumSnapshotsToKeep = 1
 
+// RollbackSamplingInterval controls how far back we are looking
+// in the history to get the rollback points.
+// For example, the default value of 1 hour ensures that the
+// protected snapshots (NumSnapshotsToKeep = 3) are:
+//    the very latest snapshot(ie the current one),
+//    the snapshot that was persisted 1 hour before the current one,
+//	  the snapshot that was persisted 2 hours before the current one
+var RollbackSamplingInterval = 1 * time.Hour
+
+// getProtectedEpochs aims to fetch the epochs keep based on a timestamp basis.
+// It tries to get NumSnapshotsToKeep snapshots, each of which are separated
+// by a time duration of RollbackSamplingInterval.
+func (s *Scorch) getProtectedEpochs(
+	persistedSnapshots []*snapshotMetaData) map[uint64]struct{} {
+
+	// make a map of epochs to protect from deletion
+	protectedEpochs := make(map[uint64]struct{}, s.numSnapshotsToKeep)
+	protectedEpochs[persistedSnapshots[0].epoch] = struct{}{}
+	nextSnapshotToProtect :=
+		persistedSnapshots[0].timeStamp.Add(s.rollbackSamplingInterval)
+	protectedSnapshots := 1
+	lastProtectedSnapshot := 0
+
+	for i := 1; i < len(persistedSnapshots); i++ {
+		// Finding that snapshot whose timestamp is just above
+		// the required timestamp, because the snapshot at the previous
+		// index is bound to be the closest possible snapshot to the
+		// required timestamp of nextSnapshotToProtect
+		if persistedSnapshots[i].timeStamp.After(nextSnapshotToProtect) {
+			protectedEpochs[persistedSnapshots[i-1].epoch] = struct{}{}
+			nextSnapshotToProtect =
+				persistedSnapshots[i-1].timeStamp.Add(s.rollbackSamplingInterval)
+			protectedSnapshots++
+			lastProtectedSnapshot = i - 1
+			if protectedSnapshots >= s.numSnapshotsToKeep {
+				return protectedEpochs
+			}
+		}
+	}
+
+	// The worst case when there aren't enough snapshots which are
+	// RollbackPointsSamplingRate apart from each other, then just
+	// protect the next s.numSnapshotsToKeep - protectedSnapshots
+	// of snapshots.
+	start := lastProtectedSnapshot + 1
+	end := start + s.numSnapshotsToKeep - protectedSnapshots
+
+	// If we don't have enough snapshots, just take all of them.
```

Comment:
Not a good idea, this'd violate the numSnapshotsToKeep contract. Also I feel the math here has been over engineered, and can be simplified.

Reply:
Over here, I've favoured having enough (`numSnapshotsToKeep`) number of snapshots over the concept of picking specific rollback points that are separated by `rollbackSamplingInterval` duration. So, if we don't have "enough" snapshots in the persisted snapshots list to guarantee that we can get `numSnapshotsToKeep` number of protected snapshots, then we just protect the next numSnapshotsToKeep - protectedSnapshots number of snapshots.

However I'm still looking out for more edge cases that need to handled, and I'll add comments for the same.